suite/tests/client-interface/drpttracer_SUDO-test.dll.c

```diff
@@ -0,0 +1,213 @@
+    dr_unregister_filter_syscall_event(event_filter_syscall);
+
+    bool ok = drmgr_unregister_thread_init_event(event_thread_init) &&
+        drmgr_unregister_thread_exit_event(event_thread_exit) &&
+        drmgr_unregister_pre_syscall_event(event_pre_syscall) &&
+        drmgr_unregister_post_syscall_event(event_post_syscall) &&
+        drmgr_unregister_tls_field(tls_idx);
+    CHECK(ok, "drmgr_unregister_*_event failed");
+
+    drmgr_exit();
+    dr_fprintf(STDERR, "all done\n");
+}
+
+static void
+event_thread_init(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)dr_thread_alloc(drcontext, sizeof(*pt));
+    memset(pt, 0, sizeof(*pt));
+    drmgr_set_tls_field(drcontext, tls_idx, (void *)pt);
+}
+
+static void
+event_thread_exit(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* If the thread's last syscall doesn't trigger post_syscall event, we need to end the
+     * tracing manually.
+     */
+    if (pt->tracer_handle != NULL) {
+        end_tracing_and_dump_trace(drcontext);
+        pt->tracer_handle = NULL;
+    }
+    dr_thread_free(drcontext, pt, sizeof(*pt));
+}
+
+static bool
+event_filter_syscall(void *drcontext, int sysnum)
+{
+    return true;
+}
+
+static bool
+event_pre_syscall(void *drcontext, int sysnum)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* If last syscall doesn't trigger post_syscall event, we need to stop its tracing
+     * here.
+     * XXX: We don't stop tracing when the application's system call returns. So we might
+     * trace some system calls called by Dynamorio's internal code.
+     */
+    if (pt->tracer_handle != NULL) {
+        end_tracing_and_dump_trace(drcontext);
+        pt->tracer_handle = NULL;
+    }
+    /* Start trace before syscall. */
+    bool ok = drpttracer_start_tracing(drcontext, DRPTTRACER_TRACING_ONLY_KERNEL, 8, 8,
+                                       &pt->tracer_handle) == DRPTTRACER_SUCCESS;
+    CHECK(ok, "drpttracer_start_tracing failed");
+    return true;
+}
+
+static void
+event_post_syscall(void *drcontext, int sysnum)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* If the syscall doesn't trigger pre_syscall event, we need to skip it. */
+    if (pt->tracer_handle == NULL) {
+        return;
+    }
+    /* End trace after syscall. */
+    end_tracing_and_dump_trace(drcontext);
+    pt->tracer_handle = NULL;
+}
+
+static void
+end_tracing_and_dump_trace(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    CHECK(pt->tracer_handle != NULL, "tracer_handle is NULL");
+    drpttracer_output_t *output;
+    bool ok = drpttracer_end_tracing(drcontext, pt->tracer_handle, &output) ==
+        DRPTTRACER_SUCCESS;
+    CHECK(ok, "drpttracer_end_tracing failed");
+    pt->recorded_syscall_num++;
+
+    /* Dump the PT trace data to threadid.syscall_id.pt. */
```

Comment:
nit: `s/threadid.syscall_id.pt/<threadid>.<syscall_id>.pt/` (I was looking for an option or something by the former name)

same below.